### ISSUE TITLE
Implement Masking augmentation (relatively high prio)

### ISSUE THREAD

**Issue by Sllambias** [Sllambias/yucca#97]
Probably based on Agave implementation - assigned you @asbjrnmunk but not for you to implement this, I just want to steal your code

**Comment by jakobamb** [comment]
Here is my code for 2D irregular image sizes (non-square):

```python
import torch

from typing import Tuple, Union


def generate_random_mask(x: torch.Tensor, mask_ratio: float, mask_patch_size: Tuple[int, int], out_type: type = int):
    # assumes x is (B, C, H, W) or (B, C, H, W, Z)

    dim = len(x.shape) - 2
    # check if x.shape is divisible by patch_size
    assert all([x.shape[i + 2] % mask_patch_size[i] == 0 for i in range(dim)]), (
        f"Incompatiple x.shape: {x.shape}" "and mask_patch_size: {mask_patch_size}"
    )

    mask = generate_1d_mask(x, mask_ratio, mask_patch_size, out_type)
    scalars = (x.shape[2] // mask_patch_size[0], x.shape[3] // mask_patch_size[1])
    return upsample_mask(mask, mask_patch_size, scalars)


def generate_1d_mask(x: torch.Tensor, mask_ratio: float, patch_size: int, out_type: type):
    assert x.shape[1] in [1, 3], "Channel dim is not 1 or 3. Are you sure?"
    assert out_type in [int, bool]
    assert len(x.shape) == 4, "Input tensor must be (B, C, H, W) // only 2D for now"

    B = x.shape[0]
    L = (x.shape[2] // patch_size[0]) * (x.shape[3] // patch_size[1])
    len_keep = int(L * (1 - mask_ratio))

    noise = torch.randn(B, L, device=x.device)

    # sort noise for each sample
    ids_shuffle = torch.argsort(noise, dim=1)
    ids_restore = torch.argsort(ids_shuffle, dim=1)

    # generate the binary mask: 0 is keep 1 is remove
    mask = torch.ones([B, L], device=x.device)
    mask[:, :len_keep] = 0
    # unshuffle to get the binary mask
    mask = torch.gather(mask, dim=1, index=ids_restore)

    if out_type == bool:
        mask = mask.bool()  # (B, H * W)
    elif out_type == int:
        mask = mask.int()

    return mask  # (B, H * W) 0 or False is keep, 1 or True is remove


def upsample_mask(
    mask: torch.Tensor,
    mask_patch_size: Union[Tuple[int, int], Tuple[int, int, int]],
    scalars: Union[Tuple[int, int], Tuple[int, int, int]],
):
    # assumes x is (B, H * W) or (B, H * W * Z)
    assert len(mask.shape) == 2
    assert all([x > 0 for x in scalars])

    dim = len(scalars)

    if dim == 2:
        mask = (
            mask.reshape(-1, scalars[0], scalars[1])
            .repeat_interleave(mask_patch_size[0], axis=1)
            .repeat_interleave(mask_patch_size[1], axis=2)
        )  # (B, H, W)
    else:
        mask = (
            mask.reshape(-1, scalars[0], scalars[1], scalars[2])
            .repeat_interleave(mask_patch_size[0], axis=1)
            .repeat_interleave(mask_patch_size[1], axis=2)
            .repeat_interleave(mask_patch_size[2], axis=3)
        )  # (B, H, W, Z)

    return mask.unsqueeze(1)  # (B, C, H, W) or (B, C, H, W, Z)
```

**Comment by asbjrnmunk** [comment]
Here is my code which supports 2 and 3d, but has constraints on image sizes: https://github.com/asbjrnmunk/agave/blob/main/src/utils/masking.py
